### PR TITLE
UCS/TOPO: distance query based on dev-path difference

### PR DESCRIPTION
**Pull request by Akshay-Venkatesh**
## What
- Use solution used for 1.6 for 1.9 as well to query distance between IB and GPU for example. 
- Tested only with cuda and IB devices
- This along with #5299 will be need in UCT/IB to penalize IB devices away from GPU for 1.9

### CODE REVIEW

File: src/ucs/sys/topo.c [hunk 1]
```diff
@@ -96,11 +97,86 @@ ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
...
+}
 
 ucs_status_t ucs_topo_get_distance(ucs_sys_device_t device1,
                                    ucs_sys_device_t device2,
                                    ucs_sys_dev_distance_t *distance)
 {
+    int score = 0;
+    int depth = 0;
+    int same  = 1;
+	int i;
```
Comment (bureddy) [inline src/ucs/sys/topo.c:118]: indentation

File: src/ucs/sys/topo.c [hunk 2]
```diff
@@ -96,11 +97,86 @@ ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
...
+        if (resolved_path1[i] != resolved_path2[i]) same = 0;
+        if (resolved_path1[i] == '/') {
+            depth++;
+            if (same) {
+                score++;
+            }
+        }
+    }
+
+    if (3 == score) {
```
Comment (bureddy) [inline src/ucs/sys/topo.c:159]: what is 3 , 4 here? can we define macro?

Comment (bureddy) [reply src/ucs/sys/topo.c:159]: can you use switch() and use single ucs_debug before return from the func

File: src/ucs/sys/topo.c (hunk 1 above)
Comment (Akshay-Venkatesh) [reply src/ucs/sys/topo.c]: done

File: src/ucs/sys/topo.c (hunk 2 above)
Comment (Akshay-Venkatesh) [reply src/ucs/sys/topo.c]: @bureddy Simplified this logic to return relative scoring instead. Can you check if it makes sense?

**Comment by yosefe** [comment]
ok to test

**Comment by Akshay-Venkatesh** [comment]
I see this:
```
/scrap/azure/agent-07/AZP_WORKSPACE/2/s/src/ucs/sys/topo.c:144:5:
index_parm: Indexing "ucs_topo_ctx.sys_dev_to_bus_lookup.bus_arr" with "device1".
```
but I'm unable to see the specific error:
```
+ echo 'not ok 1 Coverity Detected 1 failures # https://dev.azure.com/ucfconsort/ucx/_build/results?buildId=7345/ws/cov_build_devel_7345-20200630.21/output/errors/index.html'
+ echo Coverity report: 'https://dev.azure.com/ucfconsort/ucx/_build/results?buildId=7345/ws/cov_build_devel_7345-20200630.21/output/errors/index.html'
```
@bureddy Can you point me to the coverity error details?

**Comment by bureddy** [comment]
url does not work for me. 

seems the issue is a possible overrun of bus_arr[] index

```
/scrap/azure/agent-07/AZP_WORKSPACE/2/s/test/gtest/ucs/test_topo.cc:40:
overrun-call: Overrunning callee's array of size 1024 by passing argument "4294967295U" in call to "ucs_topo_get_distance".

/scrap/azure/agent-07/AZP_WORKSPACE/2/s/src/ucs/sys/topo.c:144:5:
index_parm: Indexing "ucs_topo_ctx.sys_dev_to_bus_lookup.bus_arr" with "device1".


```

File: src/ucs/sys/topo.c [hunk 3]
```diff
@@ -96,11 +97,88 @@ ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
...
+        if (sys_dev == device2) {
+            dev2_exists = 1;
+        }
+    });
+
+    if (!dev1_exists || !dev2_exists) {
+        return UCS_ERR_IO_ERROR;
+    }
+
+    bus1 = (ucs_topo_ctx.sys_dev_to_bus_lookup.bus_arr[(size_t)device1]).bus;
```
Comment (bureddy) [inline src/ucs/sys/topo.c:144]: dow we need to typecast it to size_t?

Comment (Akshay-Venkatesh) [reply src/ucs/sys/topo.c]: I tried addressing the overrun issue with an assertion instead.

**Comment by yosefe** [comment]
bot:pipe:retest

File: src/ucs/sys/topo.c [hunk 4]
```diff
@@ -96,11 +98,91 @@ ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
     return UCS_OK;
 }
 
+static void ucs_topo_get_path_with_bus(unsigned bus, char *path)
+{
+    char prefix[] = "/sys/class/pci_bus";
```
Comment (yosefe) [inline src/ucs/sys/topo.c:103]: static const char* sysfs_pci_prefix

File: src/ucs/sys/topo.c [hunk 5]
```diff
@@ -96,11 +98,91 @@ ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
...
+    int i;
+    int comp_len;
+    char path1[PATH_MAX], path2[PATH_MAX];
+    char resolved_path1[PATH_MAX], resolved_path2[PATH_MAX];
+    unsigned bus1, bus2;
+    int rp_len1, rp_len2;
+    ucs_sys_device_t sys_dev;
+
+    /* TODO: Use generic distance queury after 1.9 release but use 1.6 WAR now
+     */
```
Comment (yosefe) [inline src/ucs/sys/topo.c:130]: can remove this comment

File: src/ucs/sys/topo.c [hunk 6]
```diff
@@ -96,11 +98,91 @@ ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
...
+            dev1_exists = 1;
+        }
+        if (sys_dev == device2) {
+            dev2_exists = 1;
+        }
+    });
+
+    if (!dev1_exists || !dev2_exists) {
+        return UCS_ERR_IO_ERROR;
+    }
```
Comment (yosefe) [inline src/ucs/sys/topo.c:143]: how could we have non-existent device? we should have gotten the handle from somewhere?
just check ucs_topo_ctx.sys_dev_to_bus_lookup.bus_arr.count (and also for UCS_SYS_DEVICE_ID_UNKNOWN)

File: src/ucs/sys/topo.c [hunk 7]
```diff
@@ -96,11 +98,91 @@ ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
...
+
+    ucs_assert(device1 < UCS_TOPO_MAX_SYS_DEVICES);
+    ucs_assert(device2 < UCS_TOPO_MAX_SYS_DEVICES);
+
+    bus1 = (ucs_topo_ctx.sys_dev_to_bus_lookup.bus_arr[device1]).bus;
+    bus2 = (ucs_topo_ctx.sys_dev_to_bus_lookup.bus_arr[device2]).bus;
```
Comment (yosefe) [inline src/ucs/sys/topo.c:149]: no need for ( )

File: src/ucs/sys/topo.c [hunk 8]
```diff
@@ -96,11 +98,91 @@ ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
     return UCS_OK;
 }
 
+static void ucs_topo_get_path_with_bus(unsigned bus, char *path)
+{
+    char prefix[] = "/sys/class/pci_bus";
+
+    if (bus < 16) {
+        sprintf(path, "%s/0000:0%x", prefix, bus);
```
Comment (yosefe) [inline src/ucs/sys/topo.c:106]: use %02x fmt for bus

File: src/ucs/sys/topo.c [hunk 9]
```diff
@@ -96,11 +98,91 @@ ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
...
+
+    for (i = 0; i < comp_len; i++) {
+        if (resolved_path1[i] != resolved_path2[i]) same = 0;
+        if (resolved_path1[i] == '/') {
+            depth++;
+            if (same) {
+                score++;
+            }
+        }
+    }
```
Comment (yosefe) [inline src/ucs/sys/topo.c:177]: can we add generic function to string.{h,c}

`unsigned ucs_path_calc_distance(const char *path1, const char *path2)`

File: src/ucs/sys/topo.c [hunk 10]
```diff
@@ -96,11 +98,91 @@ ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
...
+
+    /* Use relative scoring: if depth and score match, then devices are the
+     * same; the smaller the difference, the closer distance and shorter the
+     * latency
+     */
+    ucs_debug("distance = %d", (depth - score));
+    distance->latency = 1E-7 * (depth - score);
```
Comment (yosefe) [inline src/ucs/sys/topo.c:184]: what is 1e-7 ? shall we add define?

**Comment by Akshay-Venkatesh** [comment]
@yosefe @bureddy If there are no other comments on this one, I'll port feedback changes to 1.9.x branch PR. Lmk when there is a chance.

File: src/ucs/sys/string.c [hunk 11]
```diff
@@ -293,3 +293,35 @@ const char* ucs_flags_str(char *buf, size_t max,
...
+    int rp_len1, rp_len2;
+
+    if ((NULL == realpath(path1, resolved_path1)) ||
+        (NULL == realpath(path2, resolved_path2))) {
+        return UINT_MAX;
+    }
+
+    rp_len1  = strlen(resolved_path1);
+    rp_len2  = strlen(resolved_path2);
+    comp_len = (rp_len1 <= rp_len2) ? rp_len1 : rp_len2;
```
Comment (yosefe) [inline src/ucs/sys/string.c:314]: ucs_min

File: src/ucs/sys/string.c [hunk 12]
```diff
@@ -293,3 +293,35 @@ const char* ucs_flags_str(char *buf, size_t max,
...
+
+    for (i = 0; i < comp_len; i++) {
+        if (resolved_path1[i] != resolved_path2[i]) same = 0;
+        if (resolved_path1[i] == '/') {
+            depth++;
+            if (same) {
+                score++;
+            }
+        }
```
Comment (yosefe) [inline src/ucs/sys/string.c:323]: int disnance = 0;
...
if ((resolved_path1[i] == '/)' && !same) {
   ++distance;
}
...
return distance;

File: src/ucs/sys/string.c [hunk 13]
```diff
@@ -293,3 +293,35 @@ const char* ucs_flags_str(char *buf, size_t max,
...
+        (NULL == realpath(path2, resolved_path2))) {
+        return UINT_MAX;
+    }
+
+    rp_len1  = strlen(resolved_path1);
+    rp_len2  = strlen(resolved_path2);
+    comp_len = (rp_len1 <= rp_len2) ? rp_len1 : rp_len2;
+
+    for (i = 0; i < comp_len; i++) {
+        if (resolved_path1[i] != resolved_path2[i]) same = 0;
```
Comment (yosefe) [inline src/ucs/sys/string.c:317]: code style

File: src/ucs/sys/topo.c [hunk 14]
```diff
@@ -100,88 +101,40 @@ ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
...
-            dev2_exists = 1;
-        }
-    });
-
-    if (!dev1_exists || !dev2_exists) {
+    unsigned path_distance;
+
+    if ((device1 == UCS_SYS_DEVICE_ID_UNKNOWN) ||
+        (device2 == UCS_SYS_DEVICE_ID_UNKNOWN) ||
+        (ucs_topo_ctx.sys_dev_to_bus_lookup.count < 2) ) {
```
Comment (yosefe) [inline src/ucs/sys/topo.c:119]: what if it's same device? should treat it as 0 distance

File: src/ucs/sys/string.h [hunk 15]
```diff
@@ -198,6 +198,19 @@ const char* ucs_flags_str(char *str, size_t max,
...
+ * Get estimated number of segments different in the two paths. Segments are
+ * separated by `/`.
+ *
+ * @param  path1  String pointing to first path
+ * @param  path2  String pointing to second path
+ *
+ * @return if either of the paths are invalid, UINT_MAX; if paths are the same 0
+ *         is returned; otherwise in between
+ */
+unsigned ucs_path_calc_distance(const char *path1, const char *path2);
```
Comment (yosefe) [inline src/ucs/sys/string.h:211]: return ssize_t, and UCS_ERR_xx in case of error

File: src/ucs/sys/topo.h [hunk 16]
```diff
@@ -18,6 +18,7 @@ BEGIN_C_DECLS
                                               for the device has no real bus_id
                                               E.g. virtual devices like CMA/knem
                                             */
+#define UCS_TOPO_HOP_OVERHEAD     1E-7
```
Comment (yosefe) [inline src/ucs/sys/topo.h:21]: should be in C file

File: src/ucs/sys/string.h (hunk 15 above)
Comment (Akshay-Venkatesh) [reply src/ucs/sys/string.h:211]: I've changed this but I can also change this to:
```
ucs_status_t ucs_path_calc_distance(const char *path1, const char *path2, unsigned *distance);
```

File: src/ucs/sys/topo.c (hunk 14 above)
Comment (yosefe) [reply src/ucs/sys/topo.c:120]: need to remove ucs_topo_ctx.sys_dev_to_bus_lookup.count<2 check

File: src/ucs/sys/string.c [hunk 17]
```diff
@@ -294,10 +295,9 @@ const char* ucs_flags_str(char *buf, size_t max,
     return buf;
 }
 
-unsigned ucs_path_calc_distance(const char *path1, const char *path2)
+ssize_t ucs_path_calc_distance(const char *path1, const char *path2)
 {
-    int score = 0;
-    int depth = 0;
+    int distance = 0;
```
Comment (yosefe) [inline src/ucs/sys/string.c:300]: unsigned distance

File: src/ucs/sys/topo.c [hunk 18]
```diff
@@ -130,7 +136,7 @@ ucs_status_t ucs_topo_get_distance(ucs_sys_device_t device1,
     ucs_topo_get_path_with_bus(bus2, path2);
 
     path_distance = ucs_path_calc_distance(path1, path2);
-    if (path_distance == UINT_MAX) {
+    if (path_distance < 0) {
```
Comment (yosefe) [inline src/ucs/sys/topo.c:139]: if err, convert path_distance to ucs_status_t an return it as the status, not IO_ERROR

**Comment by yosefe** [comment]
@Akshay-Venkatesh pls squash

**Comment by yosefe** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
bot:pipe:retest

**Pull request merged**